test/e2e/operator_test.go

```diff
@@ -1163,6 +1162,38 @@ func waitForAvailableReplicas(cl client.Client, ic *operatorv1.IngressController
 	return nil
 }
 
+// Wait for the provided deployment to complete its rollout.
+func waitForDeploymentComplete(cl client.Client, deployment *appsv1.Deployment, timeout time.Duration) error {
+	replicas := int32(1)
+	name := types.NamespacedName{Namespace: deployment.Namespace, Name: deployment.Name}
+	deployment = &appsv1.Deployment{}
+	err := wait.PollImmediate(1*time.Second, timeout, func() (bool, error) {
+		if err := cl.Get(context.TODO(), name, deployment); err != nil {
+			return false, nil
```

Comment:
Should `false, err` be returned here, instead of `false, nil`?

Reply:
Probably, yeah.  Generally, I think we should either `t.Logf(..., err)` or `return false, err`.  In this case, the deployment should already exist, and we're only watching for a mutation, so using `return false, err` would make sense; on the other hand, doing so would make the test more likely to break if there are transient network or API issues.  Would it make most sense in general to `t.Logf(..., err)` and `return false, nil` to decrease the likelihood that tests fail due to networking or API failures unrelated to the test?

Reply:
https://bugzilla.redhat.com/show_bug.cgi?id=1828618 is related to the general question (maybe it prompted your comment here).

Reply:
Yup, that BZ definitely prompted my comment. 

I definitely agree that we should either log the error responsibly and return `false, nil` *or* just flat out fail by returning return `false, err` in these situations.

To counter your point about transient network or API issues, if any of these issues errors were to persist for a length of time close to the timeout, returning `false, err` and failing the test immediately would be beneficial. Additionally, parts of the ingress operator's e2e test already fail on unsuccessful API calls (for example, https://github.com/openshift/cluster-ingress-operator/blob/master/test/e2e/operator_test.go#L193). So I'm wondering how much we would really be decreasing the likelihood of the tests failing from transient issues by using the responsible logging and return `false, nil` strategy. For this reason, I am leaning towards turning `false, err` and outright failing. Thoughts?

Reply:
> To counter your point about transient network or API issues, if any of these issues errors were to persist for a length of time close to the timeout, returning `false, err` and failing the test immediately would be beneficial.

I disagree—if networking or API is broken, other tests should report that breakage.  Having this component's tests fail due to other components' breakage makes it more difficult to determine which component is at fault and leads to misdirected Bugzilla reports.

> Additionally, parts of the ingress operator's e2e test already fail on unsuccessful API calls (for example, https://github.com/openshift/cluster-ingress-operator/blob/master/test/e2e/operator_test.go#L193). So I'm wondering how much we would really be decreasing the likelihood of the tests failing from transient issues by using the responsible logging and return `false, nil` strategy. For this reason, I am leaning towards turning `false, err` and outright failing. Thoughts?

That is why I am speaking generally—I believe this component's tests should not fail on errors that are caused by other components, so I am suggesting that as a general principle, we should log and retry on errors that are not caused by our component (and transient errors are very unlikely to be caused by our component).  Does that make sense?

Reply:
That does make sense, and I think you've convinced me. I agree that logging and retrying transient errors, when applicable, makes the most sense. Let's include this conversation in https://bugzilla.redhat.com/show_bug.cgi?id=1828618 and continue the conversation there. Would like to here the rest of the teams thoughts if thats OK too.